src/terrain/auth/user_attributes.clj

```diff
@@ -156,13 +156,17 @@
 
 (defn authenticate-current-user
   "Authenticates the user and binds current-user to a map that is built from the user attributes retrieved
-   during the authentication process."
+   during the authentication process. This middleware does not require authentication information to be
+   present in the request. If authentication information is present then authentication must succeed for the
+   request to be processed. Requests without credentials will be passed to the handler without authentication
+   routes that require authentication can use the require-authentication middleware."
   [handler]
   (wrap-auth-selection [[get-fake-auth           (wrap-fake-auth handler)]
                         [get-de-jwt-assertion    (wrap-de-jwt-auth handler)]
                         [get-wso2-jwt-assertion  (wrap-wso2-jwt-auth handler)]
                         [get-cas-oauth-token     (wrap-cas-oauth handler)]
-                        [get-keycloak-oidc-token (wrap-keycloak-oidc handler)]]))
+                        [get-keycloak-oidc-token (wrap-keycloak-oidc handler)]
+                        [(constantly true)       handler]]))
```

Comment:
Note that this means that the handler generated by this call to `wrap-auth-selection` will never return a 401. I briefly considered modifying `wrap-auth-selection` so that handlers generated by it never return a 401. The idea was that we'd force the caller to explicitly handle the case where there's no authentication information in the request. I reconsidered, though. If the case where authentication information isn't present in the request isn't specifically handled and we receive such a request then returning a 401 seems like the right thing to do.